[AIRFLOW-869] Refactor mark success functionality

bolkedebruin opened the pull request:
Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- AIRFLOW-869

Mark success was buggy and was directly creating and changing task instances, without considering dag runs. This patch removes direct database access from views.py and puts it in the API, making it ready for future use.  

* The recursive has been removed. It made no sense in my opinion. SubDags and their tasks are always considered. 
* It is dependent on existing dag runs. You cannot add a task instance somewhere in the middle of no-where.
* It will create dag runs for subdags to ensure consistency.

@aoen @r39132 @toddychen please provide feedback, the functionality itself should be working and testable in a local dev environment.

aoen commented on airflow/api/common/experimental/mark_tasks.py:
```diff
@@ -0,0 +1,152 @@
+
+def _create_dagruns(dag, execution_dates, state, run_id_template):
+    """
+    Infers from the dates which dag runs need to be created and does so.
```
Nit: Can add descriptions (especially for the return)

bolkedebruin replied:
Done

aoen commented on airflow/models.py:
```diff
@@ -3858,7 +3856,10 @@ def find(dag_id=None, run_id=None, execution_date=None,
             qry = qry.filter(DR.run_id == run_id)
         if execution_date:
-            qry = qry.filter(DR.execution_date == execution_date)
+            if isinstance(execution_date, list):
```
Might be worth simplifying the interface (just taking a list of execution_dates)

bolkedebruin replied:
using in_ you mean? I think the query optimiser will behave different then

aoen replied:
I meant that if execution_dates is always a list (e.g. [2016-04-10]) then you don't need this if statement or the else

bolkedebruin replied:
Ah, ok. It isn't, the function is used elsewhere where it is not a list.

aoen commented on airflow/api/common/experimental/mark_tasks.py:
```diff
@@ -0,0 +1,152 @@
+        last execution date.
+    :param past: Retroactively mark all tasks starting from start_date of the DAG
+    :param state: State to which the tasks need to be set
+    :return: list of tasks that has been created and updated
```
Nit:
s/has/have

aoen commented on airflow/api/common/experimental/mark_tasks.py:
```diff
@@ -0,0 +1,152 @@
+    :return: list of tasks that has been created and updated
+    """
+    assert isinstance(execution_date, datetime.datetime)
+    execution_date = execution_date.replace(microsecond=0)
```
Might be good adding a comment as to why this is needed.

aoen commented on airflow/api/common/experimental/mark_tasks.py:
```diff
@@ -0,0 +1,152 @@
+    dag = task.dag
+
+    # determine date range of dag runs and tasks to consider
+    end_date = ((dag.latest_execution_date or datetime.datetime.now())
```
Nit: I think this is complex enough to unwind into an if statement

aoen commented on airflow/api/common/experimental/mark_tasks.py:
```diff
@@ -0,0 +1,152 @@
+    else:
+        start_date = execution_date
+
+    start_date = execution_date if not past else start_date
```
Nit: Would be good to combine this with the if statements above to avoid overwriting previous assignments and simplifying the logic

bolkedebruin replied:
This is copy paste from the previous code, I am not sure if it would be clearer

aoen commented on tests/www/views.py:
```diff
@@ -0,0 +1,118 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unittest
```
Would be good to add a bit more coverage (e.g. for subdags), or at least a TODO

bolkedebruin replied:
Will be part of the pr

toddychen commented on airflow/www/views.py:
```diff
@@ -1095,107 +1096,22 @@ def success(self):
         recursive = request.args.get('recursive') == "true"
```
Do you want to remove the `recursive` option? I think you have mentioned it somewhere. Not only here, but also from the html and js, /www/templates/airflow/dag.html

bolkedebruin replied:
yes it doesn't make sense to me to have it as a separate option,.

toddychen reviewed (changes requested):
remove recursive option from UI and success view function

bolkedebruin commented on airflow/api/common/experimental/mark_tasks.py:
```diff
@@ -0,0 +1,152 @@
+    assert isinstance(execution_date, datetime.datetime)
+    execution_date = execution_date.replace(microsecond=0)
+
+    assert task.dag is not None
```
Doesn't assert do that ie. AssertionError? I consider its an error of the calling function if the value is not here.

aoen replied:
Fair enough

bolkedebruin committed:
[AIRFLOW-869] Refactor mark success functionality

This refactors the mark success functionality in a
more generic function that can set multiple states and
properly drills down on SubDags.
